pkg/proxy/middleware.go

```diff
@@ -0,0 +1,282 @@
+// It retries the request if the response was discarded and the response status code is retryable.
+// It uses exponential backoff for retries with a random jitter.
+// The maximum number of retries is determined by the maxRetries field of RetryMiddleware.
+func (r RetryMiddleware) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	lazyBody := NewLazyBodyCapturer(request.Body)
+	request.Body = lazyBody
+	for i := 0; ; i++ {
+		withoutRetry := i == r.maxRetries || request.Context().Err() != nil
+		if withoutRetry {
+			r.nextHandler.ServeHTTP(writer, request)
+			return
+		}
+		capturedResp := NewDiscardableResponseWriter(writer, r.isRetryableStatusCode)
+		// call next handler in chain
+		r.nextHandler.ServeHTTP(capturedResp, request.Clone(request.Context())) // clone also copies the reference to the lazy body capturer
+
+		if !r.isRetryableStatusCode(capturedResp.CapturedStatusCode()) {
+			break
+		}
+		// setup for retry
+		lazyBody.Capture()
+
+		totalRetries.Inc()
+		// exponential backoff
+		jitter := time.Duration(rand.Intn(50))
+		time.Sleep(time.Millisecond*time.Duration(1<<uint(i)) + jitter)
+	}
+}
+
+func (r RetryMiddleware) isRetryableStatusCode(status int) bool {
+	// any 1xx informational response should be written
+	if status >= 100 && status < 200 {
+		return false
+	}
+	_, ok := r.retryStatusCodes[status]
+	return ok
+}
+
+type (
+	StatusCodeCapturer interface {
+		CapturedStatusCode() int
+	}
+	XResponseWriter interface {
+		http.ResponseWriter
+		StatusCodeCapturer
+	}
+
+	// CapturedBodySource represents an interface for capturing and retrieving the body of an HTTP request.
+	CapturedBodySource interface {
+		IsCaptured() bool
+		GetBody() []byte
+	}
+
+	XBodyCapturer interface {
+		io.ReadCloser
+		CapturedBodySource
+		Capture()
+	}
+)
+
+var (
+	_ http.Flusher                    = &discardableResponseWriter{}
+	_ io.ReaderFrom                   = &discardableResponseWriterWithReaderFrom{}
+	_ CaptureStatusCodeResponseWriter = &discardableResponseWriter{}
+)
+
+// discardableResponseWriter represents a wrapper around http.ResponseWriter that provides additional
+// functionalities for handling response writing. Depending on the status code,
+// the header + content on write is skipped so that it can be re-used on retry or written to the underlying
+// response writer.
+type discardableResponseWriter struct {
+	http.ResponseWriter
+	headerBuf      http.Header
+	wroteHeader    bool
+	statusCode     int
+	immediateWrite bool
+	isDiscardable  func(status int) bool
+}
+
+// NewDiscardableResponseWriter creates a new instance of the response writer delegator.
+// It takes a http.ResponseWriter and a function to determine by the status code if content should be written or discarded (for retry).
+func NewDiscardableResponseWriter(writer http.ResponseWriter, isDiscardable func(status int) bool) XResponseWriter {
```

Comment:
I think wrapping the http.ResponseWriter comes with thorns. You can lose functionality if you don't implement all of the interfaces that ResponseWriter can be: https://go.dev/play/p/3B0yYC8kTvf

This is one the reasons I think that the middleware pattern in Go typically leads to more code than necessary. In the case of Lingo where we have a single primary handler (the proxy) I think that the middleware approach leads to more a more verbose/complicated solution than leaning on helper functions.

Reply:
Bringing benefits being customizable discussion here: https://github.com/substratusai/lingo/issues/63